crates/sui-graphql-rpc/src/types/checkpoint.rs

```diff
@@ -20,3 +24,47 @@ pub(crate) struct Checkpoint {
     // transactionConnection(first: Int, after: String, last: Int, before: String): TransactionBlockConnection
     // address_metrics: AddressMetrics,
 }
+
+impl From<&sui_json_rpc_types::Checkpoint> for Checkpoint {
```

Comment:
what's our long term story for rpc structure versioning? When i'm working on the indexer schemas, I find it tedious and error-prone to juggle between several data types.

A more concrete question: once indexer v2's schema is ready, do we still use json-rpc types such as `sui_json_rpc_types::Checkpoint` as the intermediate data structure? or do we translate the stored bytes directly to graphql structure?

Reply:
Answering the concretes first, no we don't need to use json-rpc types as the intermediate data structure if we intend to resolve fields with indexer v2. We can directly resolve data returned from indexer v2 into some graphql structure. That would involve implementing the DataProvider trait for IndexerV2 and then From<IndexerV2> for graphql types as needed. Or something like that, but no dependencies on json-rpc types

I don't quite follow your second question, but I think we'll always have some intermediate structure representing what the data provider returns before we can translate into graphql structure.

As an example, today to resolve the epoch field we need to fetch sui system state and protocol configs. Maybe indexer v2 can return this information directly. We can just update the resolver for Checkpoint.epoch to use IndexerV2DataProvider, the graphql structure stays the same even though we've modified which DataProvider it depends on. We could also use the two (and other DataProviders) in tandem to resolve fields

From what I understand though, all the json-rpc usage in this repo today is just to get a demo up, so we'll probably strip it out once indexer v2 is ready. So the intention has always been to limit the extent to which we rely on json-rpc types

I have not really thought about rpc structure versioning beyond the DataProvider trait :sweat_smile: But do you mean versioning iterations of a graphql object (like Checkpoint -> CheckpointV2) or the data providers that resolve graphql object fields?